File: python/exponential_search.py
def exponential_search(arr, n, x):
  """
  performs search on a sorted array
  arr: sorted array
  n: size of array
  x: element to be searched
  returns:
  index of element if found else -1
  """
  if arr[0] == x:
    return 0

  i = 1
  while i < n and arr[i] <= x:
    i = i * 2
  return binary_search(arr, i // 2, min(i, n - 1), x)


def binary_search(arr, left, right, x):
  if right >= left:
    mid = left + (right - left) // 2
    if arr[mid] == x:
      return mid
    if arr[mid] > x:
      return binary_search(arr, left, mid - 1, x)
    return binary_search(arr, mid + 1, right, x)
  return -1

# example usage
arr = [2, 3, 4, 10, 40]
x = 10
result = exponential_search(arr, len(arr), x)
if result == -1:
  print("element is not present in array")
else:
  print("element is not present at index", str(result))

File: c/exponentialSearch.c
#include <stdio.h>

int exponentialSearch(int arr[], int n, int x) {
  if(arr[0] == x)
    return 0;
  int i = 1;
  while(i < n && arr[i] <= x)
    i = i * 2;
  return binarySearch(arr, i/2, min(i, n - 1), x);
}

int binarySearch(int arr[], int l, int r, int x) {
  if(r >= 1) {
    int mid = l + (r - l) / 2;
    if(arr[mid] == x)
      return mid;
    if(arr[mid] > x)
      return binarySearch(arr, l, mid -1, x);
    return binarySearch(arr, mid + 1, r, x);
  }
  return -1;
}

int min(int x, int y) {
  return (x < y) ? x : y;
}

int main(void) {
  int arr[] = {2, 3, 4, 10, 40};
  int x = 10;
  int result = exponentialSearch(arr, sizeof(arr) / sizeof(arr[0]), x);
  if(result == -1)
    printf("element is not present in array \n");
  else
    printf("element is at index %d", result);
  return 0;
}
